File: LAB/Lab_5/Z1.scala
import scala.collection.immutable.LazyList

sealed trait Lazy3Tree[+A]
case object Empty extends Lazy3Tree[Nothing]

case class Node[+A](value: A, l: Lazy3Tree[A], m: Lazy3Tree[A], r: Lazy3Tree[A]) extends Lazy3Tree[A]:
  lazy val left = l
  lazy val middle = m
  lazy val right = r

def order[A](value: A, l: Lazy3Tree[A], m: Lazy3Tree[A], r: Lazy3Tree[A]): (Lazy3Tree[A], Lazy3Tree[A], Lazy3Tree[A]) =
  (l, m, r)

def traverse[A](
  orderFun: (A, Lazy3Tree[A], Lazy3Tree[A], Lazy3Tree[A]) => (Lazy3Tree[A], Lazy3Tree[A], Lazy3Tree[A]),
  tree: Lazy3Tree[A]
): LazyList[A] = tree match
  case Empty => LazyList.empty
  case Node(value, l, m, r) =>
    val (nl, nm, nr) = orderFun(value, l, m, r)
    lazy val lTail = traverse(orderFun, nl)
    lazy val mTail = traverse(orderFun, nm)
    lazy val rTail = traverse(orderFun, nr)
    LazyList.cons(value, lTail #::: mTail #::: rTail)

def generateTree(n: Int): Lazy3Tree[Int] =
  def generateTreeHelper(value: Int, depth: Int): Lazy3Tree[Int] =
    if (depth == n) Empty
    else
      lazy val l = generateTreeHelper(value * 3 + 1, depth + 1)
      lazy val m = generateTreeHelper(value * 3 + 2, depth + 1)
      lazy val r = generateTreeHelper(value * 3 + 3, depth + 1)
      Node(value, l, m, r)

  generateTreeHelper(0, 0)

@main def main() =
  val tree = generateTree(3)
  val result = traverse(order, tree)

  println(result.toList)

File: LAB/Lab_5/Z1.ml
type 'a lazy3Tree =
  | Empty
  | Node of {value: 'a; children: 'a lazy3Tree Lazy.t * 'a lazy3Tree Lazy.t * 'a lazy3Tree Lazy.t};;

type 'a lazyList =
  | Nil
  | Cons of {hd: 'a; tl: (unit -> 'a lazyList)};;

let rec (@$) l1 l2 =
  match l1 with
  | Nil -> l2
  | Cons {hd = h; tl = t} ->
      Cons{hd = h; tl = fun () -> t() @$ l2};;

let order value l m r =
  (m, l, r);; (* Czysto przykładowa implementacja, tutaj może znaleźć się cokolwiek, o ile zwraca krotkę wartości wejściowych *)

let rec traverse orderFun tree =
  match tree with
  | Empty -> Nil
  | Node {value; children = (l, m, r)} ->
    let l', m', r' = orderFun value l m r in
    let lTail = lazy (traverse orderFun (Lazy.force l'))
    and mTail = lazy (traverse orderFun (Lazy.force m'))
    and rTail = lazy (traverse orderFun (Lazy.force r')) in
    Cons {
      hd = value;
      tl = fun () -> (Lazy.force lTail) @$ (Lazy.force mTail) @$ (Lazy.force rTail)
    };;

let generateTree n =
  let rec generateTreeHelper value depth =  
    match depth with
    | j when j >= n -> Empty
    | _ ->
        let l = lazy (generateTreeHelper (value * 3 + 1) (depth + 1))
        and m = lazy (generateTreeHelper (value * 3 + 2) (depth + 1))
        and r = lazy (generateTreeHelper (value * 3 + 3) (depth + 1)) in
        Node {
          value = value;
          children = (l, m, r)
        } in
  
  generateTreeHelper 0 0

let rec lazyToRegular lazyList =
  match lazyList with
  | Nil -> []
  | Cons {hd; tl} ->
    hd :: lazyToRegular (tl());;

let tree = generateTree 3;;
let result = traverse order tree;;

lazyToRegular result;;